pkg/streamingpromql/operators/aggregations/topkbottomk/range_query.go

```diff
@@ -0,0 +1,497 @@
+	Inner                    types.InstantVectorOperator
+	Param                    types.ScalarOperator
+	TimeRange                types.QueryTimeRange
+	Grouping                 []string // If this is a 'without' aggregation, New will ensure that this slice contains __name__.
+	Without                  bool
+	MemoryConsumptionTracker *limiting.MemoryConsumptionTracker
+	IsTopK                   bool // If false, this is operator is for bottomk().
+
+	expressionPosition posrange.PositionRange
+	limit              []int64 // Maximum number of values to return at each time step for each group
+
+	remainingInnerSeriesToGroup []*rangeQueryGroup // One entry per series produced by Inner, value is the group for that series
+	remainingGroups             []*rangeQueryGroup // One entry per group, in the order we want to return them
+
+	currentGroup              *rangeQueryGroup
+	seriesIndexInCurrentGroup int
+
+	annotations                            *annotations.Annotations
+	haveEmittedIgnoredHistogramsAnnotation bool
+
+	// Reuse the same heap instance to allow us to avoid allocating a new one every time.
+	heap *rangeQueryHeap
+}
+
+var _ types.InstantVectorOperator = &RangeQuery{}
+
+func (t *RangeQuery) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadata, error) {
+	if err := t.loadLimit(ctx); err != nil {
+		return nil, err
+	}
+
+	innerSeries, err := t.Inner.SeriesMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	groups := map[string]*rangeQueryGroup{}
+	groupLabelsBytesFunc := t.groupLabelsBytesFunc()
+	t.remainingInnerSeriesToGroup = make([]*rangeQueryGroup, 0, len(innerSeries))
+
+	for seriesIdx, series := range innerSeries {
+		groupLabelsString := groupLabelsBytesFunc(series.Labels)
+		g, groupExists := groups[string(groupLabelsString)] // Important: don't extract the string(...) call here - passing it directly allows us to avoid allocating it.
+
+		if !groupExists {
+			g = &rangeQueryGroup{}
+			groups[string(groupLabelsString)] = g
+		}
+
+		g.totalSeries++
+		g.lastSeriesIndex = seriesIdx
+		t.remainingInnerSeriesToGroup = append(t.remainingInnerSeriesToGroup, g)
+	}
+
+	t.remainingGroups = make([]*rangeQueryGroup, 0, len(groups))
+
+	for _, g := range groups {
+		t.remainingGroups = append(t.remainingGroups, g)
+	}
+
+	// Sort the groups in the order we'll return them.
+	slices.SortFunc(t.remainingGroups, func(a, b *rangeQueryGroup) int {
+		return a.lastSeriesIndex - b.lastSeriesIndex
+	})
+
+	// Sort the series in the order we'll return them.
+	// It's important that we use a stable sort here, so that the order we receive series (and therefore accumulate series into their groups)
+	// matches the order we'll return those series in.
+	sort.Stable(topKBottomKOutputSorter{
+		metadata:       innerSeries,
+		seriesToGroups: slices.Clone(t.remainingInnerSeriesToGroup), // We don't want to reorder remainingInnerSeriesToGroup, but we need to keep it in sync with the list of output series during sorting.
```

Comment:
I think I understand what is happening here, but I don't understand the comment sorry :-/

Reply:
How's what I've added in fdeb024?

Reply:
Yep, got it now. Thanks